FLT/HIMExperiments/module_topology.lean

```diff
@@ -206,13 +203,159 @@ noncomputable def LinearMap.neg (M : Type*) [AddCommGroup M] [Module A M] :
+  intro i _
+  have : (fun (a : A × M) ↦ a.1 • b.equivFun a.2 i • b i) = (fun a ↦ (a.1 * b.equivFun a.2 i) • b i) := by
+    ext
+    rw [← mul_smul]
+  rw [this]
+  have : (fun a ↦ (a.1 * b.equivFun a.2 i) • b i) = (fun (a : A) ↦ a • b i) ∘ (fun (m : A × A) ↦ m.1 * m.2) ∘ (fun (m : A × M) ↦ (m.1, b.equivFun m.2 i)) := by
+    ext a
+    simp
+  rw [this]
+  apply @Continuous.comp _ _ _ (@instTopologicalSpaceProd _ _ _ (topology A)) _ (topology A) _ _
+  · have : (fun a ↦ a • b i) = (LinearMap.smul₁ A M (b i)) := by
+      unfold LinearMap.smul₁
+      simp
+    rw [this]
+    nth_rewrite 1 [Module.topology_self A]
+    apply Module.continuous_linear A (LinearMap.smul₁ A M (b i))
+  apply @Continuous.comp _ _ _ (@instTopologicalSpaceProd _ _ _ (topology A)) _ _ _ _
+  exact continuous_mul
+  apply @Continuous.prod_map _ _ _ _ _ _ _ (topology A) (fun m ↦ m) (fun m ↦ b.equivFun m i)
+  exact continuous_id
+  have : (fun m ↦ b.equivFun m i) = LinearMap.basis₁ A M ι b i := by
+    unfold LinearMap.basis₁
+    simp
+  rw [this]
+  nth_rewrite 2 [Module.topology_self A]
+  apply Module.continuous_linear A (LinearMap.basis₁ A M ι b i)
+
+
+
+-- I assume this is true! Lots of things like this seem to be true.
+lemma Module.continuous_bilinear {P : Type*} [AddCommGroup P] [Module A P] [Module.Finite A P] [Module.Free A P]
+    (f : M →ₗ[A] N →ₗ[A] P) :
+    @Continuous (M × N) P (Module.topology A) (Module.topology A) (fun mn ↦ f mn.1 mn.2) := by
+  let ι := Free.ChooseBasisIndex A M
+  let κ := Free.ChooseBasisIndex A N
+  have b : Basis ι A M := by --how can I name the index ι without doing all this?
+    simp only [ι]
+    exact Free.chooseBasis A M
+  have d : Basis κ A N := by
+    simp only [κ]
+    exact Free.chooseBasis A N
+  rw [← prod_canonical]
+  have : (fun (mn : M × N) ↦ (f mn.1) mn.2) =
+    (fun (mn : M × N) ↦  (∑ k : κ, (∑ i : ι, d.equivFun mn.2 k • b.equivFun mn.1 i • f (b i)) (d k))) := by
+    ext ⟨x, y⟩
+    simp only
+    calc
+      (f x) y = (f (∑ i : ι, b.equivFun x i • b i)) (∑ k : κ, d.equivFun y k • d k) := by
+        rw [Basis.sum_equivFun, Basis.sum_equivFun]
+      _ = ∑ k : κ, (∑ i : ι, d.equivFun y k • b.equivFun x i • f (b i)) (d k) := by simp [Finset.smul_sum]
+  rw [this]
+  apply @continuous_finset_sum _ _ _ (@instTopologicalSpaceProd _  _ (topology A) (topology A))
+    (topology A) _ _ _ Finset.univ
+  intro k _
+  have : (fun (a : M × N) ↦ (∑ i : ι, d.equivFun a.2 k • b.equivFun a.1 i • f (b i)) (d k)) =
+    fun (a : M × N) ↦ ∑ i : ι, ((d.equivFun a.2 k • b.equivFun a.1 i • f (b i)) (d k)) := by
+    simp
+  rw [this]
+  apply @continuous_finset_sum _ _ _ (@instTopologicalSpaceProd _  _ (topology A) (topology A))
+    (topology A) _ _ _ Finset.univ
+  intro i _
+  apply @Continuous.smul _ _ _ (topology A) (topology A) (@instTopologicalSpaceProd _ _ (topology A) (topology A)) _ (@Module.instContinuousSMul _ _ iA _ _ _ _ _ _ ) _ _
```

Comment:
You could do it like this:
```suggestion
  intro i _
  convert Continuous.smul ?_ ?_
  · exact Module.topology A
  · infer_instance
```
In general I think @s are a code smell, but I agree that this is pretty hairy.